### DISCUSSION TITLE
How to debug regexManagers?

### DISCUSSION THREAD

**Discussion by dbanty** [renovatebot/renovate#20881] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

In general, I want to know how to debug a regexManager. When looking at logs on https://app.renovatebot.com, I see no indication that my regexManager was run, nor any error messages about what's wrong with it.

For more info, I want to auto-update a string within my GitHub workflows:

```yaml
jobs:
  release:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0
          token: ${{ secrets.PAT }}
      - name: Install Knope
        uses: knope-dev/action@v1
        with:
          version: 0.6.2
      - run: knope release --dry-run
```

That `version: 0.6.2` defines the version of the tool which should be auto-updated. It is [a crate called knope](https://crates.io/crates/knope).

### What you tried so far.

This is my Renovate config:

```json
{
  "extends": [
    "config:base",
    ":semanticCommitTypeAll(chore)"
  ],
  "rangeStrategy": "widen",
  "regexManagers": [
    {
      "fileMatch": ["release.*\\.yml"],
      "matchStrings": ["version:\\s*(?<currentValue>.*?)"],
      "depNameTemplate": "knope",
      "datasourceTemplate": "crate"
    }
  ]
}
```

I've tried several variations on `fileMatch` and `matchStrings`, none seem to give any feedback in the Renovate logs.

### Relevant debug logs

I haven't found any relevant logs, which is the main problem. I don't even know how to diagnose 😓. Here is a gist containing the full output of one run though: https://gist.github.com/dbanty/093db02c482add9f7c2fbc89e6b61d93

**Comment by dbanty** [comment]
I fixed the specific bug—my regex was lazy. I still think it'd be worth knowing how to get more verbose info out of Renovate to find out what it's doing with the regex manager. For example, it getting a version of `""` should be an error, I'd think.

**Comment by reitzig** [comment]
+1000

In your case, the log mentions applying regex managers at [dbanty/093db02c482add9f7c2fbc89e6b61d93#file-log-txt-L331](https://gist.github.com/dbanty/093db02c482add9f7c2fbc89e6b61d93#file-log-txt-L331) and then never mentions it again.

I have the same problem with this "beauty":

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "regexManagers": [
    {
      "description": "Detect all image names of the form 'someNameImage : \"registry/image:tag@sha256:hash\"'",
      "fileMatch": ["^vars/globals.groovy"],
      "matchStrings": [
          "^\\s*\\w+Image\\s*:\\s*\"(?:(?<registryUrl>[^/]+)/)?(?<depName>[^\":@]+)(?::(?<currentValue>[^\"@]+))?(?:@(?<currentDigest>sha256:[^\"]+))?\""
      ],
      "datasourceTemplate": "docker",
      "versioningTemplate": "docker"
    }
  ]
}
```

Snippet of the file:

```groovy
def call() {
    def ecrHostname = '123456.dkr.ecr.eu-central-x.amazonaws.com'
    def dockerHubProxy = 'docker-proxy.some.org'
    vars = [
            //credentials
            bitbucketSsh                  : 'jenkins-bitbucket-ssh',
            bitbucketCreds                : 'jenkins-bitbucket-credentials',
 
            helmAgentImage                : "${ecrHostname}/shared-tools/helm-agent",
            kanikoAgentImage              : "${ecrHostname}/shared-tools/kaniko-agent",
            renovateConfigValidatorImage  : "${dockerHubProxy}/renovate/renovate:slim",
            asciidoctorAgentImage         : "${dockerHubProxy}/asciidoctor/docker-asciidoctor:latest",
            ruleTestImage                 : "docker-proxy.some.org/asciidoctor/docker-asciidoctor:latest",

            // paths
            kubeconfig                    : './',
            chartPath                     : './charts/*',
            docsTarget                    : 'docs/target',
            docsSource                    : './docs/src/main/asciidoc'
    ]

    return vars
}
```

I get, in a `lookup` dry-run on a "base" branch:

```
DEBUG: Matched 1 file(s) for manager regex: vars/globals.groovy (repository=cc/jenkins_shared_libraries, baseBranch=feat/renovate-image-refs-in-globals)
DEBUG: manager extract durations (ms) (repository=cc/jenkins_shared_libraries, baseBranch=feat/renovate-image-refs-in-globals)
       "managers": {"pre-commit": 0, "maven": 15, "regex": 15}
DEBUG: Found maven package files (repository=cc/jenkins_shared_libraries, baseBranch=feat/renovate-image-refs-in-globals)
DEBUG: Found 1 package file(s) (repository=cc/jenkins_shared_libraries, baseBranch=feat/renovate-image-refs-in-globals)
 INFO: Dependency extraction complete (repository=cc/jenkins_shared_libraries, baseBranch=feat/renovate-image-refs-in-globals)
       "stats": {
         "managers": {"maven": {"fileCount": 1, "depCount": 12}},
         "total": {"fileCount": 1, "depCount": 12}
       }
```

Clearly, the (single) regex manager is being applied to the file -- but then, silently, no dependencies are found.

Best I can determine, the regex is _syntactically_ fine and matches the relevant lines. I did the best with the matching groups and properties I could glean from the docs (I _have_ [gotten regex managers for image references to work before](https://github.com/renovatebot/renovate/issues/10993#issuecomment-1472316573)), but clearly there's a mistake. Sadly, Renovate doesn't tell us what.

FWIW, I don't expect registries of the form `${...}` to _work_ -- but I'd expect seeing 404s (or some other error) before tackling that with `depNameTemplate` or `registryUrlTemplate` (with `replace`?). I'll note that I included the fully qualified `ruleTestImage` specifically to check for this; not even that line is being picked up as expected.

**Reply by reitzig** [reply]
Uh, can't use `^` to match start-of-line:

> The regex manager matches are done per-file and not per-line, you should be aware when using the ^ and/or $ regex assertions.
> ([Source](https://docs.renovatebot.com/modules/manager/regex/#regular-expression-capture-groups))

Not sure what output I'd have needed. Something like "zero matches of X in file Y" would have sent me looking in the correct direction at least; _I_ was fiddling with `depName` and `registryUrl` ...

**Reply by reitzig** [reply]
My next stumble point was that no `eq` helper seems to be available, so

```
"matchStrings": [
    "^\\s*\\w+Image\\s*:\\s*\"(?:(?<registry>[^/]+)/)?(?<depName>[^\":@]+)(?::(?<currentValue>[^\"@]+))?(?:@(?<currentDigest>sha256:[^\"]+))?\""
]
```

with `registryUrlTemplate` wasn't intuitive to get to work. I decided to go with

```json
"matchStrings": [
    "\\n\\s*\\w+Image\\s*:\\s*\"(?<imageName>[^\":@]+)(?::(?<currentValue>[^\"@]+))?(?:@(?<currentDigest>sha256:[^\"]+))?\""
],
```

and use the `replace` helper in `depNameTemplate`. One has to somehow intuit that `$` has to be escaped in the replace string:

```json
"depNameTemplate": "{{{replace '\\${ecrHostname}' '123456.dkr.ecr.eu-central-x.amazonaws.com' (replace '\\${dockerHubProxy}' 'docker-proxy.some.org' imageName)}}}"
```

Not sure what to propose for these specific cases, but the experience is ... not good. Took me several dozen pushes and reruns to figure this out.

**Reply by reitzig** [reply]
And we're not finished, because above pattern only catches _some_ entries -- those with a tag.

Turns out, if the `currentValue` group _doesn't_ match, the whole match is silently skipped. Since for Docker giving a tag is optional and I'd still want to have the digest pinned, I tried

```json
"currentValueTemplate": "{{#if currentValue}}{{{currentValue}}}{{else}}{{/if}}"
```

but no luck. Using `latest` only gives `"skipReason": "invalid-value"` (the images are not versioned). Not sure what to do here except add the digests manually to get renovate started; this _may_ be consistent with how the `dockerfile` manager handles things as [renovatebot/renovate:lib/modules/manager/dockerfile/extract.ts#33](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/dockerfile/extract.ts#L33) seems to not do anything if neither `dep.currentValue` nor `dep.currentDigest` is  set.

**Reply by reitzig** [reply]
I'm leaving this here as an example of a debugging session that didn't feel very fruitful; no offense intended. I appreciate that Renovate does many things and that logs useful for drilling into one particular configuration would be overwhelming for a full run.

**Reply by HonkingGoose** [reply]
Hi @reitzig,

What's the problem here?

- Renovate's documentation doesn't explain how to debug regex
- Renovate itself doesn't give clear feedback about what's going wrong
- Both docs and Renovate bot itself need improvement

If you have any ideas for the documentation, we can work together on that. For any code changes, I'll let the maintainers decide. 😉

**Reply by reitzig** [reply]
Yes, all of the above. 😅

_Ideally_, there would be a tool I can run _locally_ on my repo checkout that performs the equivalent of extract and, optionally, lookup dry-runs. Having to wait for the CI system to run the dry runs is what makes all of this _super_ annoying.

As for the regex managers, I honestly don't know what to change in the docs. Sure, the how-to could be improved -- currently, relevant information is spread across multiple places. But if you read it all, you have all the relevant information. Well, had. We all know that even great docs don't replace good errors, don't we? 😉

Regarding the implementation, here are some thoughts, with little editing and in loose order:

- Align API with `packageRules`. `fileMatch` uses regex but `matchPaths` glob? They often come in matching pairs, and having to use different syntax is confusing.
- Clarify `packageName` vs `depName`. They seem to be used differently, sometimes interchangeably.
- Clarify the interface between managers and data sources. Currently, you need to know what the data source expects while configuring the manager; different sources require, admit, and interpret some values differently! (Far as I can tell, anyway.)
- Let me give each regex manager a (unique) _name_. Log output that aggregates 7 regex managers doesn't help me; I want to see what `jib-parent-image` did on `pom.xml`!
- Log levels are all over the place. 
  - INFO is useless for debugging (doesn't even contain the results of each "stage"), and DEBUG can be far too detailed (HTTP statistics?)
  - Things I consider errors for config debugging are logged under DEBUG, probably because that's our perspective for a "big" Renovate run.
  - I need a "strict"/"config development" mode that logs config errors as ERROR _and stops_.
- I need to dry-run against config _only_ from a "base" branch, without merging with the default branch.
- It would be helpful to get the JSON of config merge, extraction, lookup, and branch summary, respectively, as plain (or file) output.
  - One might even attempt to write unit tests for Renovate configs! (Which I would _love_ to do for presets. That needs more ingredients, though, once we enter `lookup` territory.)
  - Unit tests for regex managers, as one of the more complex pieces of Renovate config, would be a great start.
- `extract` should tell me for each regexManager and (matched) file what matches there were, what the match groups were, and what the values after templating were, as they would be passed to the data source.
- `extract` should report any and all problems with templating and "calling" the data source.
- `lookup` needs to log any problems; currently, we sometimes only get a short buzzword in `skipReason`.
- (in "strict" mode), `lookup` should fail whenever registry access fails. My templating may have produced something syntactically fine that is complete bogus.

**tl;dr**: I would like to have developer tooling for Renovate configs.

That's a little all over the place, I'm aware. What do you think, would any of these fly as feature requests?

**Reply by HonkingGoose** [reply]
Thank you very much for listing your problems! ❤️ 

@rarkins please read the list of "feature requests" and let me know if you want me to create issues for the feature requests you like? 😉

**Reply by HonkingGoose** [reply]
@rarkins I'm willing to break these "todos" out into their own issues, but I first need to know which todo's you'll accept issues for. 😉

**Reply by HonkingGoose** [reply]
I found this issue to allow Renovate to self-test basic functionality:

- #4403

**Reply by rarkins** [reply]
> _Ideally_, there would be a tool I can run _locally_ on my repo checkout that performs the equivalent of extract and, optionally, lookup dry-runs. Having to wait for the CI system to run the dry runs is what makes all of this _super_ annoying.

#3609 - Support local git/clone run

> As for the regex managers, I honestly don't know what to change in the docs. Sure, the how-to could be improved -- currently, relevant information is spread across multiple places. But if you read it all, you have all the relevant information. Well, had. We all know that even great docs don't replace good errors, don't we? 😉

The biggest challenge with regex managers is regex. When regex doesn't match.. it just doesn't. It's not like we can log _every single time a regex doesn't match a string in any file_ because it would be thousands or more.

> * Align API with `packageRules`. `fileMatch` uses regex but `matchPaths` glob? They often come in matching pairs, and having to use different syntax is confusing.

#21737 - Consistency of blob vs regex across different configuration options

> * Clarify `packageName` vs `depName`. They seem to be used differently, sometimes interchangeably.

We could add this to docs. For now this issue describes a little: https://github.com/renovatebot/renovate/issues/16012

> * Clarify the interface between managers and data sources. Currently, you need to know what the data source expects while configuring the manager; different sources require, admit, and interpret some values differently! (Far as I can tell, anyway.)

Needs more specifics. I suspect just that a few datasources could be better described when they're not so obvious, e.g. `git-tags`

> * Let me give each regex manager a (unique) _name_. Log output that aggregates 7 regex managers doesn't help me; I want to see what `jib-parent-image` did on `pom.xml`!

I like this idea, maybe in conjunction with us moving to the term "custom" managers to allow more than just regex.

#21760 - Allow naming of regex managers

> * Log levels are all over the place.
>   
>   * INFO is useless for debugging (doesn't even contain the results of each "stage"), 

This is intentional

> and DEBUG can be far too detailed (HTTP statistics?)

I don't agree with "far too detailed". Knowing the exact URLs called has proven useful many times for "debugging".

>   * Things I consider errors for config debugging are logged under DEBUG, probably because that's our perspective for a "big" Renovate run.

We suppress some errors into debug because we don't want to trouble administrators. But maybe we should distinguish between "user" errors versus "bot" errors so that administrators can focus on the latter only if desired.

Related: #8646 - Semantic error messages

>   * I need a "strict"/"config development" mode that logs config errors as ERROR _and stops_.

That's meant to be our default behavior (config error issue, run is aborted). 

> * I need to dry-run against config _only_ from a "base" branch, without merging with the default branch.

Please describe this in more detail. You mean for testing a new config? And if so, then wouldn't the local dry run discussed earlier be sufficient?

> * It would be helpful to get the JSON of config merge, extraction, lookup, and branch summary, respectively, as plain (or file) output.

There's already an option `printConfig` to dump the config.

`packageFiles with updates` covers extraction and lookup results.

>   * One might even attempt to write unit tests for Renovate configs! (Which I would _love_ to do for presets. That needs more ingredients, though, once we enter `lookup` territory.)
>   * Unit tests for regex managers, as one of the more complex pieces of Renovate config, would be a great start.

I don't understand what you mean by these. Within the Renovate code base, or users write their own tests against their own config?

> * `extract` should tell me for each regexManager and (matched) file what matches there were, what the match groups were, and what the values after templating were, as they would be passed to the data source.

Earlier you said debug logging is "far too detailed", but this for me would fall under that. Sounds like something you wouldn't want to log every time, and instead only during config creation. Therefore better suited to trace logging locally than debug logging everywhere

> * `extract` should report any and all problems with templating and "calling" the data source.

Need more details. Extract phase doesn't call datasource

> * `lookup` needs to log any problems; currently, we sometimes only get a short buzzword in `skipReason`.
> * (in "strict" mode), `lookup` should fail whenever registry access fails. My templating may have produced something syntactically fine that is complete bogus.

We could enable a mode where people opt into more "strict" lookups, meaning fail and log on any lookup error. You'd be surprised how many people want the opposite though (just keep going, try your best..)

Please start a new discussion on this topic (strict, more verbose logging of lookup errors), then once we have reached an agreement we can turn it into a feature request issue.

**Reply by reitzig** [reply]
Oh boy, it's been a while. My bad. Thanks for your responses!

> >  extract should tell me for each regexManager and (matched) file what matches there were, what the match groups were, and what the values after templating were, as they would be passed to the data source.
>
> Earlier you said debug logging is "far too detailed", but this for me would fall under that. Sounds like something you wouldn't want to log every time, and instead only during config creation. Therefore better suited to trace logging locally than debug logging everywhere

Right. We don't want that during production runs.  🤔 

> Extract phase doesn't call datasource

Yup. I remember needing `lookup` to reproduce some problems -- e.g. wrong use of matching groups -- but _that_ was nigh impossible to get to run locally.

I would have loved to run `lookup` _up to_ executing requests, but then just logging the URLs. At that point, I can confirm whether I fed the datasource in a wrong way.

---

Re #3609: I remember trying the new mode in the context of private infra behind VPN and authentication and immediately rejecting it:

- Shared presets didn't resolve, causing errors (IIRC). Manual workaround possible by removing the extends or copy-paste, but annoying.
- `lookup` would be tripped up by network connections to yet other protected systems (ECR registry, Nexus, ...)

_Maybe_ I could have gotten all these sorted out, but at that point I fell back to pushing N+1 times to Jenkins. :/

---

In summary, I think what I _really_ would like to have is a modified dry-runner for local usage that 

- uses the currently checked out config only (no config merge, no extends),
- runs `extract` _and_ `lookup` but
- stubs any requests to data sources with logging the request, and
- outputs otherwise excessive detail about custom managers (and data sources, for that matter).

With that tool, I would be able to iterate rapidly on custom managers using my local dev machine.

For debugging the regexes themselves, I guess a separate command of the form

```bash
renovate-regex-tester [--config renovate.json] <custom matcher name> <file>
```

that runs the exact same matching logic against exactly one file would be useful. Yes, that's basically the same as using regex101.com -- _but_ it would remove _any_ doubts about escaping, which _keeps tripping me up_. 😩

**Comment by jackson-chris** [comment]
+1 I stumbled across this after trying to debug my own regexManagers configuration and ran into a lot of the exact same issues that @reitzig has already articulated. It would be great if some of these issues that were opened up in response were prioritized.

**Reply by rarkins** [reply]
PRs welcome
